### PR TITLE
Add support for installing TLJH on Arm64 systems

### LINKED ISSUE #727: Bump traefik version
We install 1.7.18, the latest v1 version is 1.17.33 - lets bump to ensure we don't have bugs that has been resolved already.

https://github.com/jupyterhub/the-littlest-jupyterhub/blob/909b257981dee561d12ec21ea86b349a5d462ba5/tljh/traefik.py#L17

### PR DESCRIPTION
**Pull request by cdibble**
Fixes [https://github.com/jupyterhub/the-littlest-jupyterhub/issues/62#](https://github.com/jupyterhub/the-littlest-jupyterhub/issues/62#)
Fixes #727 

 - [X ] Add / update documentation
 -  This PR shouldn't change anything from the user's point of view, but I added a couple of lines to state explicitly that this supports ARM 64 and AMD64 architectures and that it uses a miniforge conda environment.
 - [ X] Add tests
 - Updated ``tests/test_conda.py`` to use the miniforge installer instead of miniconda.

### CODE REVIEW

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/tljh-on-m1-mac-arm-docker-installer-is-x86-specific/10894/2

File: tljh/utils.py [hunk 1]
```diff
@@ -51,5 +51,5 @@ def get_plugin_manager():
     pm = pluggy.PluginManager('tljh')
     pm.add_hookspecs(hooks)
     pm.load_setuptools_entrypoints('tljh')
-
+
```
Comment (consideRatio) [inline tljh/utils.py:54]:
```suggestion

```

File: tljh/traefik.py [hunk 2]
```diff
@@ -40,7 +50,7 @@ def fatal_error(e):
     giveup=fatal_error
 )
 def ensure_traefik_binary(prefix):
-    """Download and install the traefik binary"""
+    """Download and install the traefik binary-- prefix = '/opt/tljh/hub/'"""
```
Comment (consideRatio) [inline tljh/traefik.py:53]:
```suggestion
    """Download and install the traefik binary to a location identified by a prefix path such as '/opt/tljh/hub/'"""
```

File: tljh/traefik.py [hunk 3]
```diff
@@ -13,12 +13,22 @@
 from tljh.configurer import load_config, _merge_dictionaries
 
 # FIXME: support more than one platform here
-plat = "linux-amd64"
-traefik_version = "1.7.18"
+# traefik 2.7.x is not supported. Use v1.7.x
+# plat = "linux-amd64"
+# traefik_version = "1.7.18"
```
Comment (consideRatio) [inline tljh/traefik.py:18]:
```suggestion
```

File: tljh/installer.py [hunk 4]
```diff
@@ -190,8 +197,9 @@ def ensure_user_environment(user_requirements_txt_file):
     # If no prior miniconda installation is found, we can install a newer version
     else:
         logger.info('Downloading & setting up user environment...')
-        installer_url = "https://github.com/conda-forge/miniforge/releases/download/{v}/Mambaforge-{v}-Linux-x86_64.sh".format(v=mambaforge_new_version)
+        installer_url = "https://github.com/conda-forge/miniforge/releases/download/{v}/Mambaforge-{v}-Linux-{arch}.sh".format(v=mambaforge_new_version, arch=os.uname().machine)
         with conda.download_miniconda_installer(installer_url, installer_sha256) as installer_path:
+            logging.info(installer_path)
```
Comment (consideRatio) [inline tljh/installer.py:202]:
```suggestion
```

File: tljh/conda.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ def download_miniconda_installer(installer_url, sha256sum):
         os.fsync(f.fileno())
 
         if sha256_file(f.name) != sha256sum:
-            raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
+           raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
```
Comment (consideRatio) [inline tljh/conda.py:60]:
```suggestion
            raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
```

File: tests/test_conda.py [hunk 6]
```diff
@@ -24,7 +43,6 @@ def prefix():
         ])
         yield tmpdir
```
Comment (consideRatio) [inline tests/test_conda.py:45]:
```suggestion


```

File: tests/test_conda.py [hunk 7]
```diff
@@ -8,14 +8,33 @@
...
+#     miniconda_installer_sha256 = "8a324adcc9eaf1c09e22a992bb6234d91a94146840ee6b11c114ecadafc68121"
+#     installer_url = "https://repo.continuum.io/miniconda/Miniconda3-{}-Linux-x86_64.sh".format(miniconda_version)
+#     with tempfile.TemporaryDirectory() as tmpdir:
+#         with conda.download_miniconda_installer(installer_url, miniconda_installer_sha256) as installer_path:
+#             conda.install_miniconda(installer_path, tmpdir)
+#         conda.ensure_conda_packages(tmpdir, [
+#             'conda==4.8.1'
+#         ])
+#         yield tmpdir
+
```
Comment (consideRatio) [inline tests/test_conda.py:26]:
```suggestion
```

File: setup.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
     name='the-littlest-jupyterhub',
     version='0.1',
     description='A small JupyterHub distribution',
-    url='https://github.com/jupyterhub/the-littlest-jupyterhub',
+    url='https://github.com/cdibble/the-littlest-jupyterhub',
```
Comment (consideRatio) [inline setup.py:7]:
```suggestion
    url='https://github.com/jupyterhub/the-littlest-jupyterhub',
```

File: docs/topic/installer-actions.rst [hunk 9]
```diff
@@ -30,7 +30,7 @@ of `traefik <http://traefik.io/>`_. This virtual environment is completely manag
 User environment
 ================
 
-By default, a ``miniconda`` environment is installed in ``/opt/tljh/user``. This contains
+By default, a ``miniforge`` conda environment is installed in ``/opt/tljh/user``. This contains
```
Comment (consideRatio) [inline docs/topic/installer-actions.rst:33]:
```suggestion
By default, a ``mambaforge`` conda environment is installed in ``/opt/tljh/user``. This contains
```

File: docs/index.rst [hunk 10]
```diff
@@ -17,7 +17,7 @@ might still make breaking changes that have no clear upgrade pathway.
 Installation
 ============
 
-The Littlest JupyterHub (TLJH) can run on any server that is running at least
+The Littlest JupyterHub (TLJH) can run on any server (x86_64/AMD64 and aarch64/ARM64 architectures are supported) that is running at least
 **Ubuntu 18.04**. Earlier versions of Ubuntu are not supported.
```
Comment (consideRatio) [inline docs/index.rst:21]:
```suggestion
The Littlest JupyterHub (TLJH) can run on any server that is running **Ubuntu 18.04** or **Ubuntu 20.04** on a amd64 or arm64 CPU architecture. Earlier versions of Ubuntu are not supported.
```

File: bootstrap/bootstrap.py [hunk 11]
```diff
@@ -357,6 +357,13 @@ def serve_forever(server):
         logger.info('Upgrading TLJH installer...')
     run_subprocess(tljh_install_cmd)
 
+    pip_flags = ['--upgrade']
+    if os.environ.get('TLJH_BOOTSTRAP_DEV', 'no') == 'yes':
+        pip_flags.append('--editable')
+    tljh_repo_path = os.environ.get(
+        'TLJH_BOOTSTRAP_PIP_SPEC',
+        'git+https://github.com/jupyterhub/the-littlest-jupyterhub.git'
+    )
```
Comment (consideRatio) [inline bootstrap/bootstrap.py:366]:
```suggestion
```

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,149 @@
+version: 2.1
```
Comment (consideRatio) [inline .circleci/config.yml:1]:
```suggestion
# We use circleci to test our support for the Arm64 architecture. This file should mimic
# what is done in the GitHub workflows.
#
version: 2.1
```

**Review by consideRatio — CHANGES_REQUESTED** [review]
@cdibble are you okay with me cleaning up the git history of this PR a bit and applying the changes suggested in my comments?

**Comment by cdibble** [comment]
Hey @consideRatio - yes thank you for picking up the slack. I did bumble the git history badly didn't I. My apologies- a result of having rolled back a bunch of the changes in this original PR and then add some new changes from there. My own attempt to squash was ugly and I was too distracted with work projects to re-do it last week. So my apologies for that and thank you for making those changes.

Would you like me to go ahead and commit those suggestions to the PR?

**Comment by consideRatio** [comment]
@cdibble I've squashed this PR into a few commits now, does this look okay to you?

I refactored some indentation etc as well in circleci config. I have a backup available of the previous PR state, and you probably have it as well on your local computer.

**Comment by cdibble** [comment]
> @cdibble I've squashed this PR into a few commits now, does this look okay to you?
> 
> I refactored some indentation etc as well in circleci config. I have a backup available of the previous PR state, and you probably have it as well on your local computer.

LGTM. Thanks for doing that @consideRatio. You could drop that commit [08d937b](https://github.com/jupyterhub/the-littlest-jupyterhub/pull/679/commits/08d937b9902d7d947e9426fe04cf1366edbeefaa) and go back to the TLJH default of `tries=20` -- I had only bumped that up to help out with some dev and I didn't really mean for it to sneak into the PR. Of course, it doesn't really hurt anything, but in most cases where things actually work, `tries=20` is plenty.

**Comment by consideRatio** [comment]
> but in most cases where things actually work, tries=20 is plenty.

Ah I wondered about that! Okay thanks for clarifying this @cdibble, and thanks for your work with this PR, thanks for sticking with it even though we were very low on maintenance capacity for a while in this repo! :heart: :tada:

**Review by consideRatio — APPROVED** [review]

**Comment by cdibble** [comment]
Awesome @consideRatio - I am glad to be of some help and I much appreciate this project and all of its contributors!

My apologies for the added overhead here- my own need to refresh my knowledge of what was going on led to some fumbling.

**Comment by consideRatio** [comment]
Wieee! Thanks again @cdibble!!!! :heart: :tada:

**Pull request merged**